arangod/Replication/DatabaseInitialSyncer.cpp

```diff
@@ -210,130 +209,203 @@ arangodb::Result fetchRevisions(
-        // documents (that we just inserted) as documents may be replicated in
-        // unexpected order.
-        if (physical->readDocument(&trx, arangodb::LocalDocumentId(rid.id()),
-                                   mdr, arangodb::ReadOwnWrites::yes)) {
-          // already have exactly this revision no need to insert
-          break;
-        }
+          options.indexOperationMode = arangodb::IndexOperationMode::internal;
+
+          arangodb::RevisionId rid = arangodb::RevisionId::fromSlice(leaderDoc);
+          // We must see our own writes, because we may have to remove
+          if (res.ok()) {
+            sl.erase(rid);
+            ++stats.numDocsInserted;
+            break;
+          }
+
+          if (!res.is(TRI_ERROR_ARANGO_UNIQUE_CONSTRAINT_VIOLATED)) {
+            auto errorNumber = res.errorNumber();
+            res.reset(errorNumber, concatT(TRI_errno_string(errorNumber), ": ",
+                                           res.errorMessage()));
+            return res;
+          }
 
-        // remove conflict and retry
-        // errorMessage() is this case contains the conflicting key
-        auto inner = removeConflict(res.errorMessage());
-        if (inner.fail()) {
-          return res;
+          // conflicting documents (that we just inserted) as documents may be
+          // replicated in unexpected order.
+          if (physical->readDocument(&trx, arangodb::LocalDocumentId(rid.id()),
+                                     mdr, arangodb::ReadOwnWrites::yes)) {
+            // already have exactly this revision no need to insert
+            sl.erase(rid);
+            break;
+          }
+
+          // remove conflict and retry
+          // errorMessage() is this case contains the conflicting key
+          auto inner = removeConflict(res.errorMessage());
+          if (inner.fail()) {
+            return res;
+          }
         }
       }
+      // Here we expect that a lot if not all of our shopping list has come
+      // back from the leader and the shopping list is empty, however, it
+      // is possible that this is not the case, since the leader reserves
+      // the right to send fewer documents. In this case, we have to order
+      // another batch:
+      if (!sl.empty()) {
+        // Take out the nonempty list:
+        std::unordered_set<arangodb::RevisionId> newList = std::move(sl);
+        VPackBuilder requestBuilder;
+        {
+          VPackArrayBuilder list(&requestBuilder);
+          for (auto it = newList.begin(); it != newList.end(); ++it) {
+            requestBuilder.add(it->toValuePair(ridBuffer));
+          }
+        }
+
+        arangodb::network::RequestOptions reqOptions;
+        reqOptions.param("collection", leader)
+            .param("serverId", state.localServerIdString)
+            .param("batchId", std::to_string(config.batch.id));
+        reqOptions.timeout = arangodb::network::Timeout(25.0);
```

Comment:
the request should be restricted to the correct database, i.e.
```suggestion
        reqOptions.timeout = arangodb::network::Timeout(25.0);
        reqOptions.database = config.vocbase.name();      
```

Reply:
Ooops. This is important.

Reply:
Done.